airbyte-cdk/bulk/toolkits/load-gcs/src/main/kotlin/io/airbyte/cdk/load/file/gcs/GcsStreamingUpload.kt

```diff
@@ -0,0 +1,129 @@
+
+        storage.create(partBlobInfo, part)
+
+        parts[index] = partName
+
+        log.info { "Uploaded part #$index => $partName, size: ${part.size} bytes" }
+    }
+
+    /** Compose all parts into the final object. */
+    override suspend fun complete(): GcsBlob {
+        if (!isComplete.setOnce()) {
+            log.warn { "Complete called multiple times for gs://${config.gcsBucketName}/$key" }
+            return GcsBlob(key, config)
+        }
+
+        if (parts.isEmpty()) {
+            // This is not ideal but follows what S3 is doing. Ideally we would return null
+            // or throw here. But until we update the CDK to handle that use case, this is
+            // better than creating an empty object in storage.
+            log.warn { "No parts uploaded. returning a blob pointing at nothing" }
+            return GcsBlob(key, config)
+        }
+
+        // We can assume here that we won't have more than 32 parts. This should be handled
+        // in the configuration (total size / chunk_size <= 32)
+        val allPartNames = parts.values.toList()
+        check(allPartNames.size <= 32) {
+            "We are attempting to compose more than 32 parts for key $key. " +
+                "GCS is not capable of doing that."
+        }
+        log.info { "Composing ${allPartNames.size} parts into gs://${config.gcsBucketName}/$key" }
+
+        // Create the final blob with metadata
+        val finalBlobId = BlobId.of(config.gcsBucketName, key)
+        val finalBlobInfo = BlobInfo.newBuilder(finalBlobId).setMetadata(metadata).build()
+
+        // Build the compose request with all part names as sources
+        val composeRequest = ComposeRequest.newBuilder().setTarget(finalBlobInfo)
+
+        // Add each part as a source
+        parts.values.forEach { partName -> composeRequest.addSource(partName) }
+
+        // Compose it all
+        storage.compose(composeRequest.build())
+
+        // Clean up the individual part objects
+        cleanupParts()
+
+        return GcsBlob(key, config)
+    }
+
+    /**
+     * Delete all temporary part objects after successful composition. Intermediate objects get
+     * cleaned up in the multi-level composition process or after the final composition.
+     */
+    private fun cleanupParts() {
+        parts.values.forEach { partName ->
+            try {
+                storage.delete(BlobId.of(config.gcsBucketName, partName))
+                log.info { "Deleted temporary part: gs://${config.gcsBucketName}/$partName" }
+            } catch (e: Exception) {
+                log.warn(e) {
+                    "Failed to delete temporary part: gs://${config.gcsBucketName}/$partName"
+                }
+            }
+        }
+    }
+
+    private fun generateUploadId(): String {
+        val randomSuffix = (1..8).map { ('a'..'z').random() }.joinToString("")
+        return "$key-${System.currentTimeMillis()}-$randomSuffix"
```

Comment:
one thing we could do to simplify the tests, is to inject a Clock object to the constructor + replace this currentTimeMillis call. Then tests could assert on exact blob key name, instead of needing to do the `.contains` stuff

(also, it's not clear to me why we need this timestamp - isn't `key` x `randomSuffix` already enough to guarantee uniqueness? ... for that matter, shouldn't `key` already be unique?)